### PR TITLE
TreeMap: added tests for deletion of single node trees

### PR DESCRIPTION
**Pull request by mattelser**
found an opportunity for a bug in deleteMin and deleteMax where tests
would usually pass, but randomly generated certain edge cases caused
failures. All known edge cases are covered by trying to delete a
singleton with deleteMin and deleteMax, so that case has been
added to their tests.

I've added code that trips the test case to the private repo under a branch of the same name [here](https://github.com/mattelser/cs404.1/blob/treeMapTestDeletesOnSingleton/java/src/edu/berkeley/cs/util/TreeMap.java). That code will randomly pass/fail the old test and consistently fail the new test.

### CODE REVIEW

File: java/test/edu/berkeley/cs/util/TreeMap_T.java [hunk 1]
```diff
@@ -93,79 +93,92 @@ public void testPutDuplicateOverwrite() {
...
     }
 
-    Assert.assertEquals(Integer.valueOf(value), map.get(key));
+    Assert.assertEquals(new Integer(value), map.get(key));
 
     value = random.nextInt();
     map.put(key, value);
     Assert.assertTrue(map.isBST());
-    Assert.assertEquals(Integer.valueOf(value), map.get(key));
+    Assert.assertEquals(new Integer(value), map.get(key));
```
Comment (fsareshwala) [inline java/test/edu/berkeley/cs/util/TreeMap_T.java:101]: Probably want to keep these using `Integer.valueOf` -- you had made a change to move from `new Integer` and I like the use of `Integer.valueOf` a bit more -- seems more idiomatic.

File: java/test/edu/berkeley/cs/util/TreeMap_T.java [hunk 2]
```diff
@@ -93,79 +93,92 @@ public void testPutDuplicateOverwrite() {
...
 
   @Test
   public void testDeleteMin() {
     Random random = new Random();
 
-    int minimum = Integer.MAX_VALUE;
-    for (int i = 0; i < 10; i++) {
-      int r = random.nextInt();
-      map.put(r, random.nextInt());
+    int[] treeSizes = new int[]{1, 10};
```
Comment (fsareshwala) [inline java/test/edu/berkeley/cs/util/TreeMap_T.java:108]: Perhaps we should have a separate test case to test the tree size of 1? That might actually duplicate code a bit ... Or perhaps use a parameterized unit test?

File: java/test/edu/berkeley/cs/util/TreeMap_T.java [hunk 3]
```diff
@@ -93,79 +93,92 @@ public void testPutDuplicateOverwrite() {
...
 
   @Test
   public void testDeleteMax() {
     Random random = new Random();
 
-    int maximum = Integer.MIN_VALUE;
-    for (int i = 0; i < 10; i++) {
-      int r = random.nextInt();
-      map.put(r, random.nextInt());
+    int[] treeSizes = new int[]{1, 10};
```
Comment (fsareshwala) [inline java/test/edu/berkeley/cs/util/TreeMap_T.java:137]: Perhaps we should have a separate test case to test the tree size of 1? That might actually duplicate code a bit ... Or perhaps use a parameterized unit test?

File: java/test/edu/berkeley/cs/util/TreeMap_T.java [hunk 4]
```diff
@@ -93,79 +93,92 @@ public void testPutDuplicateOverwrite() {
...
   @Test
   public void testDelete() {
     Random random = new Random();
 
-    int[] keys = new int[10];
-    for (int i = 0; i < keys.length; i++) {
-      keys[i] = random.nextInt();
-      map.put(keys[i], random.nextInt());
-    }
+    int[] treeSizes = new int[]{1, 10};
```
Comment (fsareshwala) [inline java/test/edu/berkeley/cs/util/TreeMap_T.java:166]: Perhaps we should have a separate test case to test the tree size of 1? That might actually duplicate code a bit ... Or perhaps use a parameterized unit test?

File: java/test/edu/berkeley/cs/util/TreeMap_T.java (hunk 2 above)
Comment (mattelser) [reply java/test/edu/berkeley/cs/util/TreeMap_T.java]: I took a brief look at parameterizing the tests, but between the decorators making that tricky and wanting interpretation of these tests to be as plain/simple possible, I opted to just be verbose

Comment (fsareshwala) [reply java/test/edu/berkeley/cs/util/TreeMap_T.java]: Sounds good. That makes sense.
